sdk/core/azure_core/src/http/policies/request_instrumentation.rs

```diff
@@ -0,0 +1,716 @@
+    ///
+    /// # Returns
+    /// A new instance of `RequestInstrumentationPolicy`.
+    ///
+    /// # Note
+    /// This policy will only create a tracer if a tracing provider is provided in the options.
+    ///
+    /// This policy will create a tracer that can be used to instrument HTTP requests.
+    /// However this tracer is only used when the client method is NOT instrumented.
+    /// A part of the client method instrumentation sets a client-specific tracer into the
+    /// request `[Context]` which will be used instead of the tracer from this policy.
+    ///
+    pub fn new(
+        azure_namespace: Option<&'static str>,
+        crate_name: Option<&'static str>,
+        crate_version: Option<&'static str>,
+        options: &RequestInstrumentationOptions,
+    ) -> Self {
+        if let Some(tracing_provider) = &options.tracing_provider {
+            Self {
+                tracer: Some(tracing_provider.get_tracer(
+                    azure_namespace,
+                    crate_name.unwrap_or("unknown"),
+                    crate_version.unwrap_or("unknown"),
+                )),
+            }
+        } else {
+            Self { tracer: None }
+        }
+    }
+}
+
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+impl Policy for RequestInstrumentationPolicy {
+    async fn send(
+        &self,
+        ctx: &Context,
+        request: &mut Request,
+        next: &[Arc<dyn Policy>],
+    ) -> PolicyResult {
+        // If the context has a tracer (which happens when called from an instrumented method),
+        // we prefer the tracer from the context.
+        // Otherwise, we use the tracer from the policy itself.
+        // This allows for flexibility in using different tracers in different contexts.
+        let tracer = if ctx.value::<Arc<dyn crate::tracing::Tracer>>().is_some() {
+            ctx.value::<Arc<dyn crate::tracing::Tracer>>()
+        } else {
+            self.tracer.as_ref()
+        };
+
+        if let Some(tracer) = tracer {
+            let mut span_attributes = vec![
+                Attribute {
+                    key: HTTP_REQUEST_METHOD_ATTRIBUTE,
+                    value: request.method().to_string().into(),
+                },
+                Attribute {
+                    key: AZ_SCHEMA_URL_ATTRIBUTE,
+                    value: request.url().scheme().into(),
+                },
+            ];
+
+            if let Some(namespace) = tracer.namespace() {
+                // If the tracer has a namespace, we set it as an attribute.
+                span_attributes.push(Attribute {
+                    key: AZ_NAMESPACE_ATTRIBUTE,
+                    value: namespace.into(),
+                });
+            }
+
+            if !request.url().username().is_empty() || request.url().password().is_some() {
+                // If the URL contains a password, we do not log it for security reasons.
```

Comment:
We can still log the rest of URL, just omit user name and password (which we should never have on azure anyway - feel free to ignore my nitpicking)

Reply:
I'm hoping that my code *does* emit the full URL without the user name and password - there's a test case for this at the bottom of the file.

And I know it should never happen for Azure, but the user controls the endpoint being passed into the SDK - and all of the languages (except C++, which explicitly doesn't support username and passwords) will happily pass the username and password to the service. So it's theoretically possible we will get a URL with a username and password in it.